src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/SecretBuildWrapper.java

```diff
@@ -26,48 +26,95 @@
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
+import java.util.WeakHashMap;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import hudson.util.Secret;
+import org.apache.commons.codec.Charsets;
 import org.jenkinsci.plugins.credentialsbinding.MultiBinding;
 import org.kohsuke.stapler.DataBoundConstructor;
 
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
 @SuppressWarnings({"rawtypes", "unchecked"}) // inherited from BuildWrapper
 public class SecretBuildWrapper extends BuildWrapper {
 
     private /*almost final*/ List<? extends MultiBinding<?>> bindings;
 
+    private static Map<AbstractBuild<?, ?>, Collection<String>> secretsForBuild = new WeakHashMap<AbstractBuild<?, ?>, Collection<String>>();
+
+    /**
+     * Gets the {@link Pattern} for the secret values for a given build, if that build has secrets defined. If not, return
+     * null.
+     * @param build A non-null build.
+     * @return A compiled {@link Pattern} from the build's secret values, if the build has any.
+     */
+    public static @Nullable Pattern getPatternForBuild(@Nonnull AbstractBuild<?, ?> build) {
+        if (secretsForBuild.containsKey(build)) {
+            return Pattern.compile(MultiBinding.getSecretForStrings(secretsForBuild.get(build)).getPlainText());
+        } else {
+            return null;
+        }
+    }
+
     @DataBoundConstructor public SecretBuildWrapper(List<? extends MultiBinding<?>> bindings) {
         this.bindings = bindings == null ? Collections.<MultiBinding<?>>emptyList() : bindings;
     }
-    
+
     public List<? extends MultiBinding<?>> getBindings() {
         return bindings;
     }
 
-    @Override public Environment setUp(AbstractBuild build, final Launcher launcher, BuildListener listener) throws IOException, InterruptedException {
+    @Override
+    public OutputStream decorateLogger(AbstractBuild build, OutputStream logger) throws IOException, InterruptedException, Run.RunnerAbortedException {
+        if (!bindings.isEmpty()) {
+            return new Filter(build.getCharset().name()).decorateLogger(build, logger);
+        } else {
+            return logger;
+        }
+    }
+
+    @Override public Environment setUp(final AbstractBuild build, final Launcher launcher, BuildListener listener) throws IOException, InterruptedException {
         final List<MultiBinding.MultiEnvironment> m = new ArrayList<MultiBinding.MultiEnvironment>();
         for (MultiBinding binding : bindings) {
             m.add(binding.bind(build, build.getWorkspace(), launcher, listener));
         }
+
+        secretsForBuild.put(build, new HashSet<String>());
+
         return new Environment() {
             @Override public void buildEnvVars(Map<String,String> env) {
                 for (MultiBinding.MultiEnvironment e : m) {
                     env.putAll(e.getValues());
+                    secretsForBuild.get(build).addAll(e.getValues().values());
```

Comment:
Better to add this above, as soon as we have called `bind`, so we do not have to assume `buildEnvVars` has been called.